### PR TITLE
Feature/asset utilization

### PR DESCRIPTION
**Pull request by t-a-y-l-o-r**
# Description

Adds in the usage/utilization modeling and update view updates

## What changes

- added in new model to track usage
- added in view upsert side-effects to increment usage based on window constant
- usage is currently based on server time. Will do a followup to test with network time
- updated serializers as needed
- added in testing
- removed old Asset methods
- removed old Asset fields

# Pixs
Screenshot 2026-05-19 at 14 37 15

## Summary by CodeRabbit

* **New Features**
  * Asset hourly usage tracking added and exposed in API responses (weekday/hour buckets, zero-count hours omitted).

* **Changes**
  * Risk-scoring fields removed from asset representations (risk score-related attributes no longer returned).
  * Asset list endpoint optimized to avoid N+1 query behavior when loading usage data.
  * Asset upsert now records usage increments based on request time windows.

* **Tests**
  * Added comprehensive tests covering usage tracking, serialization, upsert behavior, boundary cases, and query efficiency.

[Review Change Stack](https://app.coderabbit.ai/change-stack/virtalabs/blueflow/pull/155?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: blueflow/models/asset.py [hunk 1]
```diff
@@ -383,381 +354,30 @@ def changelog_qset(self):
...
-        except ValidationError:
-            return False
-        return True
-
-    def __str__(self):
-        return f"{self.id}:{self.display_name}:{self.ip_address}"
+        hour_field = f"hour_{timestamp.hour:02d}"
+        setattr(usage, hour_field, getattr(usage, hour_field) + 1)
+        usage.last_window_started_at = window_start
+        usage.save()
```
[the remark this answers is not in the document]
Comment (t-a-y-l-o-r) [reply blueflow/models/asset.py:380]: not possible with current architecture. Will have to account for this after hackathon / when we get redis streams working

File: blueflow/models/asset.py [hunk 2]
```diff
@@ -383,381 +354,30 @@ def changelog_qset(self):
...
-    def _soft_clip(x, rmax=10, rmin=0, a=1, typ="arctan"):
-        """Limit input argument x to [0, rmax).
-
-        Domain of x is supposed to be [0, ∞).
+    def __str__(self):
+        return f"{self.id}:{self.display_name}:{self.ip_address}"
 
-        If supplied x < 0, then the output is undefined (and may raise
-        exceptions).
+    def update_usage(self, timestamp):
```
Comment (xnorkl) [inline blueflow/models/asset.py:360]: Future concern worth noting: Concurrency is not covered in the tests, eg race conditions. Shouldn't treat as a blocker for demo but worth writing tests for later. 

Checkout 'SELECT FOR UPDATE` pattern to guard against race conditions.

See 
https://docs.djangoproject.com/en/6.0/topics/testing/tools/#transactiontestcase

File: blueflow/views/asset.py [hunk 3]
```diff
@@ -220,8 +218,15 @@ def validate_open_ports_tcp(self, ports_list: list[int]) -> list[int]:
...
-        return [{} for _ in range(7)]
+    def get_usage(self, obj):
+        days = [{} for _ in range(7)]
+        for usage in obj.usage.all():
+            bucket = days[usage.day_of_week]
+            for hour in range(24):
+                count = getattr(usage, f"hour_{hour:02d}")
+                if count > 0:
+                    bucket[str(hour)] = count
+        return days
```
[the remark this answers is not in the document]
Comment (xnorkl) [reply blueflow/views/asset.py:229]: queryset = Asset.objects.all().prefetch_related('usage')

**Review by xnorkl — CHANGES_REQUESTED** [review]
I future concern and one change

File: blueflow/models/asset.py (hunk 2 above)
Comment (t-a-y-l-o-r) [reply blueflow/models/asset.py:360]: coderabbitai already brought this up ;)

**Review by xnorkl — APPROVED** [review]
